### ISSUE TITLE
`unboxedTupleNameDegree_maybe` doesn't work with GHC 9.10

### ISSUE THREAD

**Issue by RyanGlScott** [goldfirere/th-desugar#213]
If you run this program with GHC 9.8 or earlier:

```hs
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE UnboxedTuples #-}
module Main where

import Language.Haskell.TH.Desugar

main :: IO ()
main = print $ unboxedTupleNameDegree_maybe ''(#,#)
```

You will get this output, as expected:

```
$ runghc-9.8 Bug.hs
Just 2
```

With GHC 9.10, however, you will instead get:

```
$ runghc-9.10 Bug.hs
Nothing
```

Ultimately, this happens due to the fact that the [names of unboxed tuple type constructors have changed in GHC 9.10](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/8820). In GHC 9.8, we have:

```
λ> :set -XUnboxedTuples -XTemplateHaskell
λ> ''(#,#)
GHC.Prim.(#,#)
```

But in GHC 9.10, we have:

```
λ> :set -XUnboxedTuples -XTemplateHaskell
λ> ''(#,#)
GHC.Types.Tuple2#
```

The [code for parsing unboxed tuple type constructors](https://github.com/goldfirere/th-desugar/blob/a910bb140d6f9d0c69077c32f70ff08286825dff/Language/Haskell/TH/Desugar/Util.hs#L222-L230), however, assumes that only the former name will be used. We need to update this code to account for the possibility of names like `Tuple2#`.

**Comment by RyanGlScott** [comment]
We should also check for things like `Sum2#` in `unboxedSumNameDegree_maybe`.

**Comment by RyanGlScott** [comment]
Sigh... my testing in #215 wasn't thorough enough. While the example in https://github.com/goldfirere/th-desugar/issues/213#issue-2274394221 is now fixed, making a small tweak to it causes it to fail again:

```hs
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE UnboxedTuples #-}
module Main where

import Language.Haskell.TH
import Language.Haskell.TH.Desugar

main :: IO ()
main = print $ unboxedTupleNameDegree_maybe $ unboxedTupleTypeName 2
```
```
$ runghc-9.8 Bug.hs
Just 2
$ runghc-9.10 Bug.hs
Nothing
```

This one is really weird, as it turns out that `unboxedTupleTypeName 2` returns a completely different `Name` than `''(#,#)` does:

```
λ> unboxedTupleTypeName 2
GHC.Prim.Tuple2#
λ> ''(#,#)
GHC.Types.Tuple2#
```

See [GHC#24750](https://gitlab.haskell.org/ghc/ghc/-/issues/24750). This is unfortunate from a `th-desugar` perspective because we [desugar uses of `UnboxedTupleT <N>` to `ConT  (unboxedTupleTypeName <N>)`](https://github.com/goldfirere/th-desugar/blob/5aa3585028b05b90e78ead54e4f5fda6ace13ac2/Language/Haskell/TH/Desugar/Core.hs#L1186), so it is particularly important to get the `unboxedTupleTypeName` case correct.

My hope is that we can fix GHC#24750 upstream so that this example Just Works™. If not, we'll need to add a special case for the (non-existent) `GHC.Prim.Tuple<N>#` family of `Name`s.

**Comment by RyanGlScott** [comment]
The fix for [GHC#24750](https://gitlab.haskell.org/ghc/ghc/-/issues/24750) is present in GHC 9.10.1, which means that the program in https://github.com/goldfirere/th-desugar/issues/213#issuecomment-2090264027 now works as expected. I've added that program (among others) as regression tests in https://github.com/goldfirere/th-desugar/commit/81793635f365bac6c5eed408aeb2e777ae9ee847.
